api/web3server.go

```diff
@@ -0,0 +1,870 @@
+}
+
+func (svr *Web3Server) isSyncing() (interface{}, error) {
+	return false, nil
+}
+
+func (svr *Web3Server) isMining() (interface{}, error) {
+	return false, nil
+}
+
+func (svr *Web3Server) getHashrate() (interface{}, error) {
+	return "0x500000", nil
+}
+
+func (svr *Web3Server) getBlockTransactionCountByHash(in interface{}) (interface{}, error) {
+	h, err := getStringFromArray(in, 0)
+	if err != nil {
+		return nil, err
+	}
+	blkMeta, err := svr.coreService.BlockMetaByHash(removeHexPrefix(h))
+	if err != nil {
+		// wrap err not found
+		return nil, err
+	}
+	return uint64ToHex(uint64(blkMeta.NumActions)), nil
+}
+
+func (svr *Web3Server) getBlockByHash(in interface{}) (interface{}, error) {
+	h, isDetailed, err := getStringAndBoolFromArray(in)
+	if err != nil {
+		return nil, err
+	}
+	blkMeta, err := svr.coreService.BlockMetaByHash(removeHexPrefix(h))
+	if err != nil {
+		// wrap err not found
+		return nil, err
+	}
+	return svr.getBlockWithTransactions(blkMeta, isDetailed)
+}
+
+func (svr *Web3Server) getTransactionByHash(in interface{}) (interface{}, error) {
+	h, err := getStringFromArray(in, 0)
+	if err != nil {
+		return nil, err
+	}
+	actionInfos, err := svr.coreService.Action(removeHexPrefix(h), true)
+	if err != nil {
+		return nil, err
+	}
+	return svr.getTransactionCreateFromActionInfo(actionInfos)
+}
+
+func (svr *Web3Server) getLogs(filter *filterObject) (interface{}, error) {
+	from, to, err := svr.parseBlockRange(filter.FromBlock, filter.ToBlock)
+	if err != nil {
+		return nil, err
+	}
+	return svr.getLogsWithFilter(from, to, filter.Address, filter.Topics)
+}
+
+func (svr *Web3Server) getTransactionReceipt(in interface{}) (interface{}, error) {
+	// parse action hash from request
+	actHashStr, err := getStringFromArray(in, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// acquire action receipt by action hash
+	actHash, err := hash.HexStringToHash256(removeHexPrefix(actHashStr))
+	if err != nil {
+		return nil, errors.Wrapf(errUnkownType, "actHash: %s", actHashStr)
+	}
+	receipt, blkHash, err := svr.coreService.ReceiptByAction(actHash)
+	if err != nil {
+		return nil, err
+	}
+
+	// read contract address from receipt
+	var contractAddr *string
+	if len(receipt.ContractAddress) > 0 {
+		res, err := ioAddrToEthAddr(receipt.ContractAddress)
+		if err != nil {
+			return nil, err
+		}
+		contractAddr = &res
+	}
+
+	// acquire transaction index by action hash
+	actInfo, err := svr.coreService.Action(removeHexPrefix(actHashStr), true)
+	if err != nil {
+		return nil, err
+	}
+	tx, err := svr.getTransactionFromActionInfo(actInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse logs from receipt
+	var logs []logsObject
+	for _, v := range receipt.Logs() {
+		var topics []string
+		for _, tpc := range v.Topics {
+			topics = append(topics, "0x"+hex.EncodeToString(tpc[:]))
+		}
+		addr := ""
+		if contractAddr != nil {
+			addr = *contractAddr
+		}
+		log := logsObject{
+			BlockHash:        "0x" + blkHash,
```

Comment:
so here return expects to have 0x at beginning?

Reply:
yes